scripts/inference.py

```diff
@@ -3,67 +3,53 @@
-    list_logprob = []
-    for i, sample in enumerate(samples_tok):
-        valid_length = len(sample) - 1  # Last token doesn't have a next token
-        sample_logprobs = logprobs[i, :valid_length, :]  # [valid_length, vocab_size]
-
-        # Extract the probabilities of the actual next tokens
-        next_tokens = sample[
-            1 : valid_length + 1
-        ]  # Tokens that follow each token in the sequence
-        correct_logprobs = sample_logprobs[torch.arange(valid_length), next_tokens]
-
-        list_logprob.append(correct_logprobs)
-    return list_logprob
+from delphi.eval.utils import get_all_and_next_logprobs, load_validation_dataset
 
-    # outputs a list of lists of correct token LOG probabilities.
-    # correct_logprobs = get_correct_logprobs(model, val_sequences[:10])
+torch.set_grad_enabled(False)
 
 
-def main(model_name, dataset_split, batch_size):
-    val_ds = load_dataset(
-        "delphi-suite/tinystories-v2-clean-tokenized", split=dataset_split
-    )
-    # val_ds[0]["tokens"]    # access first sample
+def main(model_name: str, batch_size: Int, dataset_name: str, token: str):
+    """
+    Outputs the log probabilities of the next token for each token in the validation dataset.
+    And uploads the resulting dataset to huggingface.
+    Args:
+    - model_name: The name of the model to use for inference
+    - batch_size: The batch size for processing. 80 worked well in CPU.
+    - dataset_name: The name of the dataset from which validation set will be loaded
+    - token: Hugging Face API token
+    """
+    val_ds = load_validation_dataset(dataset_name)
 
     # model accepts 2D tensors (batch_size, seq_len)
     val_sequences = torch.tensor([s["tokens"] for s in val_ds])
-
-    output_folder = "Correct_logprobs"
-    os.makedirs(output_folder, exist_ok=True)
-
-    # Initialize an empty DataFrame to accumulate log probabilities
-    accumulated_df = pd.DataFrame()
-
     model = AutoModelForCausalLM.from_pretrained(model_name)
 
-    # Loop over the validation dataset in batches
+    accumulated_logprobs = torch.tensor([], dtype=torch.float32)
+
     for i in tqdm(range(0, len(val_sequences), batch_size)):
         batch_sequences = val_sequences[i : i + batch_size]
-        batch_logprobs = get_correct_logprobs(model, batch_sequences)
-        # Convert batch log probabilities to a DataFrame
-        batch_df = pd.DataFrame([logprob.tolist() for logprob in batch_logprobs])
-        # Append the batch DataFrame to the accumulated DataFrame
-        accumulated_df = pd.concat([accumulated_df, batch_df], ignore_index=True)
-
-    # Save the accumulated DataFrame to a Parquet file
-    output_file = os.path.join(output_folder, f'{model_name.replace("/", "-")}.parquet')
-    accumulated_df.to_parquet(output_file)
+        _, next_logprobs = get_all_and_next_logprobs(model, val_sequences)
+        accumulated_logprobs = torch.cat((accumulated_logprobs, next_logprobs), dim=0)
```

Comment:
This approach is pretty inefficient, you copy all the data accumulated so far in each step. Instead you could append next_logprobs to `logprobs_list` and then at the end have `all_logprobs = torch.cat(logprobs_list, dim=0)`.

Reply:
But actually, you're converting it all to list anyway to make a DataFrame. But then Pandas stores it's data as numpy arrays 🙈 It's fine for now, but please remind me on 1-1 to chat about this.